HW2/P2/parallel_vector.pyx

```diff
@@ -71,20 +79,41 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
                              np.int32_t[:] dest,
                              int repeat):
    cdef:
-       int idx, r
+       int idx, r, ind1, ind2
        omp_lock_t *locks = get_N_locks(counts.shape[0])
 
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for each element of counts to parallelize
-   # data movement.  Be sure to avoid deadlock, and double-locking.
-   ##########
+   # &locks is an array of 1000 omp locks, with each lock corresponding to each position in the count array
+   # note: counts begins as a numpy array with index = value
+   # note: src and dest represent arrays of random indices of the counts array
+   
    with nogil:
+       # execute the following script 100 times (repeat = 100)
        for r in range(repeat):
-           for idx in range(src.shape[0]):
+           # loop through each index in src (and dest; total = 1,000,000) - use four threads here
+           for idx in prange(src.shape[0], num_threads=4):
+               # need to order locks to avoid deadlocking
+               if src[idx] > dest[idx]:
+                   ind1 = dest[idx]
+                   ind2 = src[idx]
+               else:
+                   ind1 = src[idx]
+                   ind2 = dest[idx]
+
+               # acquire locks accordingly
+               acquire(&locks[ind1])
+	       # check for double-locking
+               if ind1 != ind2:
```

Comment:
Cool way to check for checking for double locking once again, I did something slightly different. It's always fun to see different ways of implementing the same thing!